### PR TITLE
Fix frang on  h2 requests

### LINKED ISSUE #1477: Host header processing in HTTP/2
## Scope

This is follow up on https://github.com/tempesta-tech/tempesta/pull/1458#discussion_r536306942 : 

http/2 logic is not covered yet: although h2 specification doesn't forbid `host` header, an pseudo header `:authority` _should_ be used  by client, and `host` header must be ignored if `:authority` is provided. 

Usual clients, like curl don't send `host` header and prefer `:authority`, as stated in the RFC. But our frang only checks `host` header  for `http_host_required`  limits. Thus the whole logic of `http_host_required` doesn't work for h2 protocol.

## Testing

Some h2/ test is required.

### PR DESCRIPTION
**Pull request by ikoveshnikov**
Fix #1477

### CODE REVIEW

**Review by krizhanovsky — APPROVED** [review]
I didn't find any major issues, but some more work is required.

File: tempesta_fw/http.c [hunk 1]
```diff
@@ -5310,6 +5318,10 @@ tfw_http_req_process(TfwConn *conn, TfwStream *stream, const TfwFsmData *data)
 	 * for every new request. Connection rates limits usually much more
 	 * strict than request rates, so this attack path is closed by usual
 	 * frang configuration.
+	 *
+	 * TODO: block the request if Frang didn't finish its job. It mostly
+	 * indicates internal errors, when not all frang checks was asserted
+	 * against the request.
```
[the remark this answers is not in the document]
Comment (ikoveshnikov) [reply tempesta_fw/http.c:5324]: The comment mostly talked about previous issues (exist on master branch), where frang never reached DONE sate for h2 requests.  I.e. frang has several major stages: headers partially received, headers fully received, body partly received, all the frang check passed; but on h2 request processing frang got stuck in 'headers partially received' and none of the further checks wasn't called. 

At this lines a request is fully received and parsed, so if the frang has never reached `done` state, some security checks was missed. We need either check last frang state here or guarantee that frang never returns TFW_PASS until request is fully parsed.

[the remark this answers is not in the document]
Comment (krizhanovsky) [reply tempesta_fw/http.c:5324]: Could you please adjust the comment in the way it becomes precise in what the issues are and what must be done, preferably with the links to GitHub issues.

File: tempesta_fw/http.c [hunk 2]
```diff
@@ -5202,6 +5204,12 @@ tfw_http_req_process(TfwConn *conn, TfwStream *stream, const TfwFsmData *data)
 			}
 		}
 		else {
+			if (TFW_MSG_H2(req)) {
```
[the remark this answers is not in the document]
Comment (ikoveshnikov) [reply tempesta_fw/http.c]: The code was reworked.

File: tempesta_fw/http_limits.c [hunk 3]
```diff
@@ -632,58 +702,88 @@ static int
...
+		val_name = &prim_name;
+		val_trim = &prim_trim;
+		if (unlikely(TFW_STR_EMPTY(val_trim))) {
+			frang_msg("Request authority is unknown",
+				  &FRANG_ACC2CLI(ra)->addr, "\n");
+			return TFW_BLOCK;
+		}
+		else if (frang_assert_host_header(&prim_trim, &sec_trim)) {
+			frang_msg("Request authority differs between headers",
+				  &FRANG_ACC2CLI(ra)->addr, "\n");
```
[the remark this answers is not in the document]
Comment (ikoveshnikov) [reply tempesta_fw/http_limits.c:763]: Done, the message is HTTP/1.1 specific now, functional tests was updated to reflect the change. But I used the same messages for purpose: the situations are the same for both protocols and it's easier for administrator to grep log for security events without protocol-specific information.

File: tempesta_fw/http_limits.c [hunk 4]
```diff
@@ -632,58 +702,88 @@ static int
...
+			return TFW_BLOCK;
+		}
+		break;
+	/*
+	 * Old protocols may have no 'host' header, it's arbitrary header that
+	 * can be treated by modern servers as special host headers, while
+	 * old ones may not. Client tries to confuse some servers in the
+	 * forwarding chain by adding a header with possible double meaning?
```
[the remark this answers is not in the document]
Comment (ikoveshnikov) [reply tempesta_fw/http_limits.c]: The comment was reworded. Hope it's more clear now.

File: tempesta_fw/http_parser.c [hunk 5]
```diff
@@ -5292,6 +5300,107 @@ do {									\
...
+	__FSM_DECLARE_VARS(req);
+
+	__FSM_START(parser->_i_st);
+
+	__FSM_STATE(Req_I_Authority) {
+		if (!H2_MSG_VERIFY(TFW_HTTP_HDR_H2_AUTHORITY))
+			return CSTR_NEQ;
+		__set_bit(TFW_HTTP_B_URI_FULL, req->flags);
+		__FSM_I_JMP(Req_I_A_Start);
+	}
```
[the remark this answers is not in the document]
Comment (ikoveshnikov) [reply tempesta_fw/http_parser.c]: Dummy states were removed, function duplicate was removed.

File: tempesta_fw/http.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (ikoveshnikov) [reply tempesta_fw/http.c]: Issue number was added into TODO comment. The corresponding issue has a full description of the problem now.

**Comment by ikoveshnikov** [comment]
> Please update <wiki>

Done.

**Review by avbelov23 — APPROVED** [review]

**Review by krizhanovsky — APPROVED** [review]
LGTM

**Pull request merged**